applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c

```diff
@@ -0,0 +1,786 @@
+		[LWM2M_CARRIER_ERROR_FOTA_CONN_LOST] =
+			"Connection to remote server lost",
+		[LWM2M_CARRIER_ERROR_FOTA_FAIL] =
+			"Modem firmware update failed",
+		[LWM2M_CARRIER_ERROR_CONFIGURATION] =
+			"Illegal object configuration detected",
+		[LWM2M_CARRIER_ERROR_INIT] =
+			"Initialization failure",
+		[LWM2M_CARRIER_ERROR_INTERNAL] =
+			"Internal failure",
+		[LWM2M_CARRIER_ERROR_RUN] =
+			"Configuration failure",
+	};
+
+	__ASSERT(PART_OF_ARRAY(strerr[err->type]),
+		 "Unhandled liblwm2m_carrier error");
+
+	LOG_ERR("LWM2M_CARRIER_EVENT_ERROR: %s, reason %d", strerr[err->type], err->value);
+}
+
+void print_deferred(const lwm2m_carrier_event_t *evt)
+{
+	const lwm2m_carrier_event_deferred_t *def = evt->data.deferred;
+
+	static const char * const strdef[] = {
+		[LWM2M_CARRIER_DEFERRED_NO_REASON] =
+			"No reason given",
+		[LWM2M_CARRIER_DEFERRED_PDN_ACTIVATE] =
+			"Failed to activate PDN",
+		[LWM2M_CARRIER_DEFERRED_BOOTSTRAP_NO_ROUTE] =
+			"No route to bootstrap server",
+		[LWM2M_CARRIER_DEFERRED_BOOTSTRAP_CONNECT] =
+			"Failed to connect to bootstrap server",
+		[LWM2M_CARRIER_DEFERRED_BOOTSTRAP_SEQUENCE] =
+			"Bootstrap sequence not completed",
+		[LWM2M_CARRIER_DEFERRED_SERVER_NO_ROUTE] =
+			"No route to server",
+		[LWM2M_CARRIER_DEFERRED_SERVER_CONNECT] =
+			"Failed to connect to server",
+		[LWM2M_CARRIER_DEFERRED_SERVER_REGISTRATION] =
+			"Server registration sequence not completed",
+		[LWM2M_CARRIER_DEFERRED_SERVICE_UNAVAILABLE] =
+			"Server in maintenance mode",
+		[LWM2M_CARRIER_DEFERRED_SIM_MSISDN] =
+			"Waiting for SIM MSISDN",
+	};
+
+	__ASSERT(PART_OF_ARRAY(strdef[def->reason]),
+		 "Unhandled deferred reason");
+
+	LOG_INF("LWM2M_CARRIER_EVENT_DEFERRED: reason %s, timeout %d seconds",
+		strdef[def->reason], def->timeout);
+}
+
+int lwm2m_carrier_event_handler(const lwm2m_carrier_event_t *event)
+{
+	int err = 0;
+
+	switch (event->type) {
+	case LWM2M_CARRIER_EVENT_INIT:
+		LOG_DBG("LWM2M_CARRIER_EVENT_INIT");
+		err = lte_lc_init();
+		if (err != 0) {
+			LOG_ERR("Failed to initialize lte link controller library");
+		}
+		lte_lc_register_handler(lte_event_handler);
+		break;
+	case LWM2M_CARRIER_EVENT_LTE_LINK_UP:
+		LOG_DBG("LWM2M_CARRIER_EVENT_LTE_LINK_UP");
+		err = lte_lc_connect_async(NULL);
+		break;
+	case LWM2M_CARRIER_EVENT_LTE_LINK_DOWN:
+		LOG_DBG("LWM2M_CARRIER_EVENT_LTE_LINK_DOWN");
+		err = lte_lc_offline();
+		break;
+	case LWM2M_CARRIER_EVENT_LTE_POWER_OFF:
+		LOG_DBG("LWM2M_CARRIER_EVENT_LTE_POWER_OFF");
+		err = lte_lc_power_off();
+		break;
+	case LWM2M_CARRIER_EVENT_BOOTSTRAPPED:
+		LOG_DBG("LWM2M_CARRIER_EVENT_BOOTSTRAPPED");
+		break;
+	case LWM2M_CARRIER_EVENT_REGISTERED:
+		LOG_DBG("LWM2M_CARRIER_EVENT_REGISTERED");
+		break;
+	case LWM2M_CARRIER_EVENT_DEFERRED:
+		print_deferred(event);
+		break;
+	case LWM2M_CARRIER_EVENT_FOTA_START:
+		LOG_DBG("LWM2M_CARRIER_EVENT_FOTA_START");
+#if !defined(CONFIG_SLM_NATIVE_TLS)
+		err = do_secure_socket_close_all();
```

Comment:
Assumption here. SLM has opened sockets here and there, not just those in Socket service. Some are done in NCS/Zephyr libraries, e.g. HTTP. Please consider asking MCU to shutdown all active services

Reply:
This code has now been removed, the FOTA_START event is given to the app and so is the responsibility to close the sockets upon receiving it.